src/middlewared/middlewared/plugins/cluster_linux/ctdb_public.py

```diff
@@ -0,0 +1,238 @@
+                verrors.add(
+                    f'{schema_name}.{data["ip"]}',
+                    f'"{data["ip"]}" does not exist in the cluster.'
+                )
+
+        verrors.check()
+
+    @private
+    def update_file(self, data, verrors, delete=False):
+        """
+        Update the ctdb cluster public IP file.
+        """
+
+        ctdb_file = pathlib.Path(CTDBConfig.GM_PUB_IP_FILE.value)
+        etc_file = pathlib.Path(CTDBConfig.ETC_PUB_IP_FILE.value)
+
+        # make sure the public ip file exists
+        try:
+            ctdb_file.touch(exist_ok=True)
+        except Exception as e:
+            raise CallError(f'Failed creating {ctdb_file} with error {e}')
+
+        # we need to make sure that the local etc public ip file
+        # is symlinked to the ctdb shared volume public ip file
+        symlink_it = delete_it = False
+        if etc_file.exists():
+            if not etc_file.is_symlink():
+                # delete it since we're symlinking it
+                delete_it = True
+            else:
+                # means it's a symlink but not to the ctdb
+                # shared volume public ip file
+                if not etc_file.resolve() == ctdb_file:
+                    delete_it = True
+        else:
+            symlink_it = True
+
+        if delete_it:
+            try:
+                etc_file.unlink()
+            except Exception as e:
+                raise CallError(f'Failed deleting {etc_file} with errror {e}')
+
+        if symlink_it:
+            try:
+                etc_file.symlink_to(ctdb_file)
+            except Exception as e:
+                raise CallError(f'Failed symlinking {etc_file} to {ctdb_file} with error {e}')
+
+        if not delete:
+            entry = f'{data["ip"]}/{data["netmask"]} {data["interface"]}'
+            with open(PUB_IP_FILE, 'a') as f:
+                f.write(entry + '\n')
+        else:
+            with open(PUB_IP_FILE) as f:
+                lines = f.read().splitlines()
+
+            lines.remove(next(i for i in lines if data['ip'] in i))
+            with open(PUB_IP_FILE, 'w') as f:
+                f.writelines(map(lambda x: x + '\n', lines))
+
+    @accepts(Dict(
+        'node_create',
+        IPAddr('ip', required=True),
+        Int('netmask', required=True),
+        Str('interface', required=True),
+    ))
+    @job(lock=PUB_LOCK)
+    async def do_create(self, job, data):
+        """
+        Add a ctdb public address to the cluster
+
+        `ip` is an IP v4/v6 address
+        `netmask` is a cidr notated netmask (i.e. 16/24/48/64 etc)
+        `interface` a network interface to have assigned the `ip`
+        """
+
+        schema_name = 'node_create'
+        verrors = ValidationErrors()
+
+        await self.middleware.call('ctdb.public.ips.common_validation', data, schema_name, verrors)
+        await self.middleware.call('ctdb.public.ips.update_file', data, verrors)
+
+        return data
```

Comment:
May need to add the `id` to the returned dictionary. Will probably have to go through self.query() to ensure that id is consistent.